app/controllers/SearchController.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Controllers;
+
+use App\Services\CacheService;
+use Memcached;
+
+class SearchController extends BaseController
+{
+    public CacheService $searchService;
+    private Memcached $memcache;
+    public array $data;
+
+    public function __construct()
+    {
+        $this->memcache = new Memcached();
```

Comment:
$this->proxySearchService = new SearchServiceProxy(new SearchService);

Reply:
final readonly class ProxySearchService implements SearchServiceInterface
{
public function __construct(
    private SearchServiceInterface $searchService, 
) {
$this->memcache->addServer('localhost', 11211);
}

public function search(string $query): array
{
     $cacheKey = self::getCacheKey($query);

     if (! $this->memcache->exists($cacheKey)) {
         $value = $this->searchService->search($query);

         $this->memcache->store($cacheKey, $value);
     }
     
     return $this->memcache->get($cacheKey));
}

private static function getCacheKey(string $query): string
{
    return sprintf('ProxySearchService:%s', $query);
}
}